index/ask/runner/runner.go

```diff
@@ -0,0 +1,299 @@
+			}
+		}
+	}
+}
+
+// update triggers a full-scan generates and saves a new fresh index and builds
+// views for better querying.
+func (ai *Runner) update() error {
+	log.Info("updating ask index...")
+	defer log.Info("ask index updated")
+	startTime := time.Now()
+	newIndex, cache, err := generateIndex(ai.ctx, ai.api)
+	if err != nil {
+		return fmt.Errorf("generating index: %s", err)
+	}
+
+	ai.lock.Lock()
+	ai.index = newIndex
+	ai.orderedAsks = cache
+	ai.lock.Unlock()
+
+	stats.Record(context.Background(), metrics.MFullRefreshDuration.M(time.Since(startTime).Milliseconds()))
+	ai.signaler.Signal()
+
+	return nil
+}
+
+// generateIndex returns a fresh index
+func generateIndex(ctx context.Context, api *apistruct.FullNodeStruct) (ask.Index, []*ask.StorageAsk, error) {
+	addrs, err := api.StateListMiners(ctx, types.EmptyTSK)
+	if err != nil {
+		return ask.Index{}, nil, err
+	}
+
+	rateLim := make(chan struct{}, qaRatelim)
+	var lock sync.Mutex
+	newAsks := make(map[string]ask.StorageAsk)
+	for i, addr := range addrs {
+		if ctx.Err() != nil {
+			break
+		}
+		rateLim <- struct{}{}
+		go func(addr address.Address) {
+			defer func() { <-rateLim }()
+			sask, ok, err := getMinerStorageAsk(ctx, api, addr)
+			if err != nil {
+				log.Errorf("getting miner storage ask: %s", err)
+				return
+			}
+			if !ok {
+				return
+			}
+			lock.Lock()
+			newAsks[addr.String()] = sask
+			lock.Unlock()
+		}(addr)
+		if i%100 == 0 {
+			stats.Record(context.Background(), metrics.MFullRefreshProgress.M(float64(i)/float64(len(addrs))))
+			log.Debugf("progress %d/%d", i, len(addrs))
+		}
+	}
+	for i := 0; i < qaRatelim; i++ {
+		rateLim <- struct{}{}
+	}
+
+	if ctx.Err() != nil {
+		return ask.Index{}, nil, fmt.Errorf("refresh was canceled")
+	}
+
+	stats.Record(context.Background(), metrics.MFullRefreshProgress.M(1))
+	ctx, _ = tag.New(context.Background(), tag.Insert(metrics.TagAskStatus, "FAIL"))
+	stats.Record(ctx, metrics.MAskQueryResult.M(int64(len(addrs)-len(newAsks))))
+	ctx, _ = tag.New(context.Background(), tag.Insert(metrics.TagAskStatus, "OK"))
+	stats.Record(ctx, metrics.MAskQueryResult.M(int64(len(newAsks))))
+
+	cache := generateOrderedAsks(newAsks)
+	return ask.Index{
+		LastUpdated:        time.Now(),
+		StorageMedianPrice: calculateMedian(cache),
+		Storage:            newAsks,
+	}, cache, nil
+}
+
+// getMinerStorage ask returns the result of querying the miner for its current Storage Ask.
+// If the miner has zero power, it won't be queried returning false.
+func getMinerStorageAsk(ctx context.Context, api *apistruct.FullNodeStruct, addr address.Address) (ask.StorageAsk, bool, error) {
+	ctx, cancel := context.WithTimeout(ctx, qaTimeout)
+	defer cancel()
+	power, err := api.StateMinerPower(ctx, addr, types.EmptyTSK)
+	if err != nil {
+		return ask.StorageAsk{}, false, fmt.Errorf("getting power %s: %s", addr, err)
+	}
+	if power.MinerPower.RawBytePower.IsZero() {
+		return ask.StorageAsk{}, false, nil
+	}
```

Comment:
If the miner power is zero, quit. Avoid doing work on useless miners.
Currently, there're >100k registered miners and only less than 50 have power.
That means saving ~100k API calls to Lotus and query-ask timeouts at the network level.

Reply:
wow, huge difference!